--- a/fs/namei.c	2018-12-01 11:30:07.672594412 -0500
+++ b/fs/namei.c	2018-12-01 11:30:58.772816410 -0500
@@ -902,8 +902,8 @@ static inline void put_link(struct namei
 		path_put(&last->link);
 }
 
-int sysctl_protected_symlinks __read_mostly = 0;
-int sysctl_protected_hardlinks __read_mostly = 0;
+int sysctl_protected_symlinks __read_mostly = 1;
+int sysctl_protected_hardlinks __read_mostly = 1;
 int sysctl_protected_fifos __read_mostly;
 int sysctl_protected_regular __read_mostly;
 
